### ISSUE TITLE
GradCAM Gradients all zeros

### ISSUE THREAD

**Issue by RaphaelMeudec** [sicara/tf-explain#54]
`GradientTape` in the GradCAM method often outputs only zeros (in `examples/callbacks/mnist.py`), resulting in None heatmaps. This has to be investigated.

![Capture d’écran de 2019-07-29 17-20-38](https://user-images.githubusercontent.com/12402673/62060397-591b3480-b225-11e9-9770-e32a93efd293.png)

**Comment by RaphaelMeudec** [comment]
Solution to this is removing the softmax layer of the grad cam model with `grad_model.layers[-1].activation = None`. However, this has side effects on all the callbacks as models are not duplicated.

**Comment by matheushent** [comment]
@RaphaelMeudec I think a solution would be swap the softmax activation to linear activation. I'm working on this but not getting success 'till now

`grad_model.layers[-1].activation = tf.keras.activations.linear`

Something like this would work, but note that just doing it doesn't change the graph, and that is the part I'm struggling

**Comment by matheushent** [comment]
The problem was solved to me swapping the softmax layer after compiling it and then calling the GradCAM class.

`def apply_modifications(model, custom_objects=None):`
       `model_path = os.path.join(tempfile.gettempdir(), next(tempfile._get_candidate_names()) + '.h5')`
       `try`:
             `model.save(model_path)`
             `return load_model(model_path, custom_objects=custom_objects)`
       `finally:`
              `os.remove(model_path)`

`if __name__ == "__main__":`
    `model = tf.keras.models.load_model('cifar10_logs/cifar10_model.hdf5')`
`    model.compile(`
`       loss='categorical_crossentropy',`
`        optimizer=tf.keras.optimizers.RMSprop(learning_rate=0.0001, decay=1e-6),`
`        metrics=['accuracy']`
`    )`

`    model.layers[-1].activation = tf.keras.activations.linear`
`    model = apply_modifications(model)`

**Comment by RaphaelMeudec** [comment]
Thanks @matheushent for the investigation. I'm not sure whether tf-explain should or shouldn't apply modifications to input models. I'll think about it, and leaving this open so people can have a reference!

**Comment by ArtDoctor** [comment]
Thanks for the solution. Is there any way to keep the output of the last layer, while still making GradCAM non-zero?
I am trying to work with my own visualization, and I wanted to keep the output prediction and the heatmap in the same place, so that it takes less time.
